### PR TITLE
Fixed #26902 -- Added require_https argument to is_safe_url()

### PR DESCRIPTION
**Pull request by suligap**
And use the argument in Django's views where it makes sense, i.e.
`django.contrib.auth` and `django.views.i18n`

Fixes https://code.djangoproject.com/ticket/26902

### CODE REVIEW

File: django/utils/http.py [hunk 1]
```diff
@@ -315,8 +319,9 @@ def _is_safe_url(url, host):
     # URL and might consider the URL as scheme relative.
     if unicodedata.category(url[0])[0] == 'C':
         return False
+    valid_schemes = ['https'] if secure else ['http', 'https']
```
Comment (andrewnester) [inline django/utils/http.py]: the only problem with check I guess that urls like `//testserver/` will be considered as secure which is not I guess

File: tests/utils_tests/test_http.py [hunk 2]
```diff
@@ -139,6 +139,10 @@ def test_is_safe_url(self):
         self.assertTrue(http.is_safe_url('/confirm/me@example.com'))
         # Basic auth without host is not allowed.
         self.assertFalse(http.is_safe_url(r'http://testserver\@example.com'))
+        # HTTPS only allowed when secure=True.
+        self.assertFalse(http.is_safe_url('http://example.com', 'example.com', secure=True))
+        # HTTP allowed by default.
+        self.assertTrue(http.is_safe_url('http://example.com', 'example.com'))
```
Comment (andrewnester) [inline tests/utils_tests/test_http.py]: could you please more test cases to check if url secured or not, please see my example of test case here https://github.com/django/django/pull/6928/files#diff-01a1ff7510f29ab8ea151390573f5736R143

File: django/utils/http.py (hunk 1 above)
Comment (suligap) [inline django/utils/http.py]: Thank you, very good point. This is now fixed.

File: django/utils/http.py [hunk 3]
```diff
@@ -315,8 +319,15 @@ def _is_safe_url(url, host):
     # URL and might consider the URL as scheme relative.
     if unicodedata.category(url[0])[0] == 'C':
         return False
+    scheme = url_info.scheme
+    # If we have a URL like //example.com/p - it doesn't have a scheme but does
+    # have a hostname, consider the scheme to be http for the purposes of
+    # further scheme validation.
+    if not url_info.scheme and url_info.netloc:
+        scheme = 'http'
```
Comment (berkerpeksag) [inline django/utils/http.py]: We can make this a one-liner:

```py
scheme = 'http' if not url_info.scheme and url_info.netloc else url_info.scheme
```

File: tests/utils_tests/test_http.py [hunk 4]
```diff
@@ -140,6 +140,24 @@ def test_is_safe_url(self):
         # Basic auth without host is not allowed.
         self.assertFalse(http.is_safe_url(r'http://testserver\@example.com'))
 
+    def test_is_safe_url_secure_param(self):
+        secure_urls = (
+            'https://example.com/p',
+            'HTTPS://example.com/p',
+            '/view/?param=http://example.com',
+        )
+        unsecure_urls = (
```
Comment (berkerpeksag) [inline tests/utils_tests/test_http.py]: It would be better to move this into a separate test.

File: tests/utils_tests/test_http.py [hunk 5]
```diff
@@ -140,6 +140,24 @@ def test_is_safe_url(self):
...
+            'http://example.com/p',
+            'ftp://example.com/p',
+            '//example.com/p',
+        )
+        for url in secure_urls:
+            self.assertTrue(http.is_safe_url(url, 'example.com', secure=True))
+        for url in unsecure_urls:
+            self.assertFalse(http.is_safe_url(url, 'example.com', secure=True))
+        # HTTP still allowed by default.
+        self.assertTrue(http.is_safe_url('http://example.com/p', 'example.com'))
```
Comment (berkerpeksag) [inline tests/utils_tests/test_http.py]: It looks like this case is already tested in `test_is_safe_url`.

File: django/utils/http.py [hunk 6]
```diff
@@ -277,12 +277,15 @@ def is_same_domain(host, pattern):
...
-def is_safe_url(url, host=None):
+def is_safe_url(url, host=None, secure=False):
     """
     Return ``True`` if the url is a safe redirection (i.e. it doesn't point to
     a different host and uses a safe scheme).
 
     Always returns ``False`` on an empty url.
+
+    If ``secure`` argument is ``True``, only https will be considered as a
+    valid scheme as opposed to both http and https with default ``False``.
```
Comment (berkerpeksag) [inline django/utils/http.py]: http -> HTTP
https -> HTTPS

File: tests/utils_tests/test_http.py (hunk 4 above)
Comment (berkerpeksag) [inline tests/utils_tests/test_http.py]: unsecure -> insecure (I'm not a native speaker so don't trust me on this :))

File: django/utils/http.py (hunk 6 above)
Comment (berkerpeksag) [inline django/utils/http.py]: We also need to add a release note for the new argument in `docs/releases/1.11.txt`.

File: django/utils/http.py (hunk 3 above)
Comment (suligap) [inline django/utils/http.py]: Sure, thanks. I just think it's more readable this way.

**Comment by suligap** [comment]
Thanks @berkerpeksag for the review. I've added the release note and other minor changes you mentioned.

File: docs/releases/1.11.txt [hunk 7]
```diff
@@ -280,6 +280,13 @@ URLs
 
 * ...
 
+Utils
+~~~~~
+
+* ``utils.http.is_safe_url()`` accepts a new ``secure`` argument. If
```
Comment (berkerpeksag) [inline docs/releases/1.11.txt]: I think the "Miscellaneous" section would be a better place to put this in.

**Comment by berkerpeksag** [comment]
A minor comment about the commit log: Could you also replace "Add" with "Added"?

File: docs/releases/1.11.txt [hunk 8]
```diff
@@ -399,6 +399,10 @@ Miscellaneous
 * The :djadmin:`makemessages` command now requires configured settings, like
   most other commands.
 
+* ``utils.http.is_safe_url()`` accepts a new ``secure`` argument. If
```
Comment (timgraham) [inline docs/releases/1.11.txt]: I'm not sure this needs to be mentioned in backwards-incompatible changes changes? Maybe Django's usage of the new argument could be considered backwards-incompatible but that's not mentioned. By the way, I'd make Django's usage of the new argument in a separate commit. You can use "Refs #26902 --" for that commit message.

File: django/utils/http.py [hunk 9]
```diff
@@ -277,12 +277,15 @@ def is_same_domain(host, pattern):
...
 
-def is_safe_url(url, host=None):
+def is_safe_url(url, host=None, secure=False):
     """
     Return ``True`` if the url is a safe redirection (i.e. it doesn't point to
     a different host and uses a safe scheme).
 
     Always returns ``False`` on an empty url.
+
+    If ``secure`` argument is ``True``, only HTTPS will be considered as a
```
Comment (timgraham) [inline django/utils/http.py]: chop "argument" or say "the secure argument"

File: django/utils/http.py [hunk 10]
```diff
@@ -277,12 +277,15 @@ def is_same_domain(host, pattern):
...
+    valid scheme as opposed to both HTTP and HTTPS with default ``False``.
```
Comment (timgraham) [inline django/utils/http.py]: the default

File: tests/auth_tests/test_views.py [hunk 11]
```diff
@@ -551,6 +551,23 @@ def test_security_check(self, password='password'):
             self.assertEqual(response.status_code, 302)
             self.assertIn(good_url, response.url, "%s should be allowed" % good_url)
 
+    def test_security_check_https(self, password='password'):
```
Comment (timgraham) [inline tests/auth_tests/test_views.py]: No need for a `password` parameter. I'll remove that current pattern... not sure how it got started.

File: django/utils/http.py [hunk 12]
```diff
@@ -315,8 +319,15 @@ def _is_safe_url(url, host):
     # URL and might consider the URL as scheme relative.
     if unicodedata.category(url[0])[0] == 'C':
         return False
+    scheme = url_info.scheme
+    # If we have a URL like //example.com/p - it doesn't have a scheme but does
```
Comment (timgraham) [inline django/utils/http.py]: more simply: `Consider URLs without a scheme (e.g. //example.com/p) to be http.`

**Comment by timgraham** [comment]
I wonder if something like `require_secure` or `require_https` might be a more explanatory parameter name?

File: docs/releases/1.11.txt (hunk 8 above)
Comment (suligap) [inline docs/releases/1.11.txt]: > I'm not sure this needs to be mentioned in backwards-incompatible changes changes?

Of course not. That wasn't intentional :) Thanks, fixed now!

And usage in django views is moved to separate commits and is also mentioned in release notes.

**Comment by suligap** [comment]
> I wonder if something like require_secure or require_https might be a more explanatory parameter name?

Absolutely! `require_https` sounds better.

Thanks for the review @timgraham

**Comment by timgraham** [comment]
merged in 5e5a17028f4b9cfb5ff777d8c259e079bca0c988, 549b90fab33c80d1ba6575d4837ea52d79f70fb1, and 1f68bb5683608ebe1ce47f0c9d37dad9482d1df9, thanks!
